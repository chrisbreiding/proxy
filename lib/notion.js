const rp = require('request-promise')

const getData = ({ notionPageId, notionToken }) => {
  return rp({
    uri: `https://api.notion.com/v1/blocks/${notionPageId}/children`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2021-05-13',
    },
    json: true,
  })
}

module.exports = {
  getData,
}
