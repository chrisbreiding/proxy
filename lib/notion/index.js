const { addUpcomingWeek } = require('./upcoming-week')
const { getData } = require('./data')
const { onSocket } = require('./shopping')

module.exports = {
  addUpcomingWeek,
  getData,
  onSocket,
}
