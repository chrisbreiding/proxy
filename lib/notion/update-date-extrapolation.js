const dayjs = require('dayjs')
const debug = require('debug')('proxy:scripts')

const {
  getBlockChildrenDeep,
  getPlainText,
  makeTextPart,
  updateBlock,
} = require('./util')
const { compact, mapPromisesSerially } = require('../util/collections')
const { getEnv } = require('../util/env')

const getCategoryBlocks = ({ notionToken, dateExtrapolationId }) => {
  return getBlockChildrenDeep({
    notionToken,
    pageId: dateExtrapolationId,
    includeId: true,
    filter: (block) => block.type === 'toggle',
  })
}

const getDateFromBlock = (block) => {
  const text = (getPlainText(block) || '').trim()
  const date = dayjs(text)

  return date.isValid() ? date : undefined
}

const getHistoricalDates = (blocks) => {
  return blocks.reduce((memo, block) => {
    const text = (getPlainText(block) || '').trim()

    if (!['Historical', 'Recent'].includes(text)) return memo

    const dates = block[block.type].children.map(getDateFromBlock)

    dates.sort((a, b) => a - b)

    return memo.concat(compact(dates))
  }, [])
}

const getIntervals = (dates) => {
  return dates.reduce((memo, date, index) => {
    const nextDate = dates[index + 1]

    if (!nextDate) return memo

    return memo.concat(nextDate.diff(date, 'day'))
  }, [])
}

const getSum = (nums) => {
  return nums.reduce((sum, num) => sum + num, 0)
}

const getAverage = (nums) => {
  return Math.round(getSum(nums) / nums.length)
}

const getExtrapolatedDateBlocks = (blocks) => {
  const block = blocks.find((block) => {
    const text = getPlainText(block) || ''

    return text.includes('Extrapolated')
  })

  return block[block.type].children.filter((b) => b.type === 'bulleted_list_item')
}

const getExtrapolatedDates = (historicalDates, blocks) => {
  const extrapolatedDateBlocks = getExtrapolatedDateBlocks(blocks)
  const intervals = getIntervals(historicalDates)
  const averageInterval = getAverage(intervals)
  const lastHistoricalDate = historicalDates[historicalDates.length - 1]

  return extrapolatedDateBlocks.reduce((memo, block) => {
    const date = memo.lastDate.add(averageInterval, 'days')

    return {
      lastDate: date,
      dates: memo.dates.concat({ date, block }),
    }
  }, { lastDate: lastHistoricalDate, dates: [] }).dates
}

const updateDates = async ({ extrapolatedDates, notionToken }) => {
  return mapPromisesSerially(extrapolatedDates, ({ date, block }) => {
    const content = {
      bulleted_list_item: {
        rich_text: [makeTextPart(date.format('YYYY-MM-DD'))],
      },
    }

    return updateBlock({ notionToken, block: content, blockId: block.id })
  })
}

const updateDateExtrapolation = async ({ notionToken, dateExtrapolationId }) => {
  const blocks = await getCategoryBlocks({ notionToken, dateExtrapolationId })
  const historicalDates = getHistoricalDates(blocks)
  const extrapolatedDates = getExtrapolatedDates(historicalDates, blocks)

  await updateDates({ extrapolatedDates, notionToken })
}

const main = async () => {
  const notionToken = getEnv('NOTION_SARAH_TOKEN')
  const dateExtrapolationId = getEnv('NOTION_DATE_EXTRAPOLATION_ID')

  debug('ENV:', {
    notionToken,
    dateExtrapolationId,
  })

  try {
    // eslint-disable-next-line no-console
    console.log('Updating date extrapolation...')

    await updateDateExtrapolation({ notionToken, dateExtrapolationId })

    // eslint-disable-next-line no-console
    console.log('Successfully updated date extrapolation')
  } catch (error) {
    // eslint-disable-next-line no-console
    console.log('Updating date extrapolation failed:')
    // eslint-disable-next-line no-console
    console.log(error.stack)
  }
}

main.updateDateExtrapolation = updateDateExtrapolation

module.exports = main
