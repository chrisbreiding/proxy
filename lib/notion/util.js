const rp = require('request-promise')

const makeRequest = ({ notionToken, body, method = 'get', path }) => {
  return rp({
    method,
    body,
    uri: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2021-05-13',
    },
    json: true,
  })
}

const getBlockChildren = ({ notionToken, pageId }) => {
  return makeRequest({ notionToken, path: `blocks/${pageId}/children` })
}

const appendBlockChildren = ({ notionToken, pageId, blocks }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${pageId}/children`,
    body: {
      children: blocks,
    },
  })
}

const updateBlock = ({ notionToken, blockId, block }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${blockId}`,
    body: block,
  })
}

const getPlainText = (withText) => {
  if (!withText || !withText.text) return

  return withText.text
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}

const weatherIconMap = {
  'clear-day': '☀️',
  'clear-night': '☀️',
  'rain': '☔️',
  'snow': '❄️',
  'sleet': '🌨',
  'wind': '💨',
  'fog': '🌫',
  'cloudy': '☁️',
  'partly-cloudy-day': '⛅️',
  'partly-cloudy-night': '⛅️',
  'default': '🌑',
}

const getWeatherIcon = (iconName) => {
  return weatherIconMap[iconName] || weatherIconMap.default
}

const makeTextPart = (content, color) => {
  const textPart = { text: { content } }

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

module.exports = {
  getBlockChildren,
  appendBlockChildren,
  updateBlock,
  getPlainText,
  getWeatherIcon,
  makeTextPart,
}
