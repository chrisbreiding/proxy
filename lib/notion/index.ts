import { addUpcomingWeek, upcomingWeekView } from './upcoming-week'
import { getBlockChildren } from './util'
import { onSocket } from './shopping'

interface GetDataOptions {
  notionToken: string
  notionPageId: string
}

export function getNotionData ({ notionToken, notionPageId }: GetDataOptions) {
  return getBlockChildren({ notionToken, pageId: notionPageId })
}

export {
  addUpcomingWeek,
  upcomingWeekView,
  onSocket,
}
