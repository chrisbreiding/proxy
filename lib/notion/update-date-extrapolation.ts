import dayjs from 'dayjs'
import Debug from 'debug'

import {
  Block,
  getBlockChildrenDeep,
  getBlockPlainText,
  makeTextPart,
  NotionBlock,
  OwnBlock,
  updateBlock,
} from './util'
import { compact, mapPromisesSerially } from '../util/collections'
import { getEnv } from '../util/env'

const debug = Debug('proxy:scripts')

interface GetCategoryBlocksOptions {
  notionToken: string
  dateExtrapolationId: string
}

function getCategoryBlocks ({ notionToken, dateExtrapolationId }: GetCategoryBlocksOptions) {
  return getBlockChildrenDeep({
    notionToken,
    pageId: dateExtrapolationId,
    includeId: true,
    filter: (block) => block.type === 'toggle',
  })
}

function getDateFromBlock (block: Block) {
  const text = (getBlockPlainText(block) || '').trim()
  const date = dayjs(text)

  return date.isValid() ? date : undefined
}

function getHistoricalDates (blocks: NotionBlock[]) {
  return blocks.reduce((memo, block) => {
    const text = (getBlockPlainText(block) || '').trim()

    if (!['Historical', 'Recent'].includes(text)) return memo

    if (!block.children) return memo

    const maybeDates = block.children.map(getDateFromBlock)
    const dates = compact(maybeDates) as dayjs.Dayjs[]

    dates.sort((a, b) => a.valueOf() - b.valueOf())

    return memo.concat(dates)
  }, [] as dayjs.Dayjs[])
}

function getIntervals (dates: dayjs.Dayjs[]) {
  return dates.reduce((memo, date, index) => {
    const nextDate = dates[index + 1]

    if (!nextDate) return memo

    return memo.concat(nextDate.diff(date, 'day'))
  }, [] as number[])
}

function getSum (nums: number[]) {
  return nums.reduce((sum, num) => sum + num, 0)
}

function getAverage (nums: number[]) {
  return Math.round(getSum(nums) / nums.length)
}

function getExtrapolatedDateBlocks (blocks: NotionBlock[]) {
  const block = blocks.find((block) => {
    const text = getBlockPlainText(block) || ''

    return text.includes('Extrapolated')
  })

  if (!block || !block.children) {
    return []
  }

  return block.children.filter((child) => {
    return child.type === 'bulleted_list_item'
  }) as NotionBlock[]
}

interface DatesAndBlocks {
  date: dayjs.Dayjs
  block: NotionBlock
}

function getExtrapolatedDates (historicalDates: dayjs.Dayjs[], blocks: NotionBlock[]) {
  const extrapolatedDateBlocks = getExtrapolatedDateBlocks(blocks)
  const intervals = getIntervals(historicalDates)
  const averageInterval = getAverage(intervals)
  const lastHistoricalDate = historicalDates[historicalDates.length - 1]

  return extrapolatedDateBlocks.reduce((memo, block) => {
    const date = memo.lastDate.add(averageInterval, 'days')

    return {
      lastDate: date,
      dates: memo.dates.concat({ date, block }),
    }
  }, { lastDate: lastHistoricalDate, dates: [] as DatesAndBlocks[] }).dates
}

interface UpdateDatesOptions {
  extrapolatedDates: DatesAndBlocks[]
  notionToken: string
}

async function updateDates ({ extrapolatedDates, notionToken }: UpdateDatesOptions) {
  return mapPromisesSerially(extrapolatedDates, ({ date, block }) => {
    const content = {
      type: 'bulleted_list_item' as const,
      content: {
        rich_text: [makeTextPart(date.format('YYYY-MM-DD'))],
      },
    } as OwnBlock

    return updateBlock({ notionToken, block: content, blockId: block.id! })
  })
}

interface UpdateDateExtrapolationOptions {
  notionToken: string
  dateExtrapolationId: string
}

export async function updateDateExtrapolation ({ notionToken, dateExtrapolationId }: UpdateDateExtrapolationOptions) {
  const blocks = await getCategoryBlocks({ notionToken, dateExtrapolationId })
  const historicalDates = getHistoricalDates(blocks)
  const extrapolatedDates = getExtrapolatedDates(historicalDates, blocks)

  await updateDates({ extrapolatedDates, notionToken })
}

export default async function main () {
  const notionToken = getEnv('NOTION_SARAH_TOKEN')!
  const dateExtrapolationId = getEnv('NOTION_DATE_EXTRAPOLATION_ID')!

  debug('ENV:', {
    notionToken,
    dateExtrapolationId,
  })

  try {
    // eslint-disable-next-line no-console
    console.log('Updating date extrapolation...')

    await updateDateExtrapolation({ notionToken, dateExtrapolationId })

    // eslint-disable-next-line no-console
    console.log('Successfully updated date extrapolation')
  } catch (error: any) {
    // eslint-disable-next-line no-console
    console.log('Updating date extrapolation failed:')
    // eslint-disable-next-line no-console
    console.log(error?.stack || error)
  }
}
