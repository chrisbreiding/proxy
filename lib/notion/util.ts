import type { Method } from 'axios'
import dayjs from 'dayjs'
import type {
  ApiColor,
  BlockObjectResponse,
  BookmarkBlockObjectResponse,
  BulletedListItemBlockObjectResponse,
  CalloutBlockObjectResponse,
  ChildDatabaseBlockObjectResponse,
  ChildPageBlockObjectResponse,
  ColumnBlockObjectResponse,
  ColumnListBlockObjectResponse,
  DividerBlockObjectResponse,
  EmbedBlockObjectResponse,
  EquationBlockObjectResponse,
  FileBlockObjectResponse,
  Heading1BlockObjectResponse,
  Heading2BlockObjectResponse,
  Heading3BlockObjectResponse,
  ImageBlockObjectResponse,
  LinkPreviewBlockObjectResponse,
  LinkToPageBlockObjectResponse,
  ListBlockChildrenResponse,
  ListDatabasesResponse,
  NumberedListItemBlockObjectResponse,
  ParagraphBlockObjectResponse,
  PdfBlockObjectResponse,
  QuoteBlockObjectResponse,
  RichTextItemResponse,
  SyncedBlockBlockObjectResponse,
  TableBlockObjectResponse,
  TableOfContentsBlockObjectResponse,
  TableRowBlockObjectResponse,
  TemplateBlockObjectResponse,
  ToDoBlockObjectResponse,
  ToggleBlockObjectResponse,
  UnsupportedBlockObjectResponse,
  UpdateBlockBodyParameters,
  VideoBlockObjectResponse,
} from '@notionhq/client/build/src/api-endpoints'

import { clone, compact } from '../util/collections'
import { debug } from '../util/debug'
import { request } from '../util/network'

export const dateRegex = /(?:Sun|Mon|Tue|Wed|Thu|Fri|Sat), (\d{1,2}\/\d{1,2})/

export function getDateFromText (dateText: string) {
  const currentDate = dayjs()
  // originally assume the date's year matches the current year
  const date = dayjs(`${dateText}/${currentDate.year()}`, 'M/D/YYYY')
  // if the current month is after the date's month, we've crossed over years
  // and the date's year should be the next year. usually, it's because it's
  // currently December, but the date's month is January. this can be
  // generically applied, so it will work if it's currently June, but we're
  // trying to get the weather for April. we don't want past weather, only
  // future weather, so assume it's for the following year.
  if (currentDate.month() > date.month()) {
    return date.add(1, 'year')
  }

  return date
}

interface MakeRequestOptions {
  body?: object
  method?: Method
  notionToken: string
  path: string
}

function makeRequest<T> (options: MakeRequestOptions): Promise<T> {
  const { notionToken, body, path, method = 'get' } = options

  return request({
    method,
    body,
    url: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2022-06-28',
    },
  })
}

interface GetBlockChildrenOptions {
  notionToken: string
  pageId: string
}

export async function getBlockChildren ({ notionToken, pageId }: GetBlockChildrenOptions) {
  const { results } = await makeRequest<ListBlockChildrenResponse>({ notionToken, path: `blocks/${pageId}/children` }) as { results: BlockObjectResponse[] }

  return results.map(convertNotionBlockToOwnBlock)
}

interface GetBlockChildrenDeepOptions {
  filter?: (block: NotionBlock) => boolean
  includeId?: boolean
  notionToken: string
  pageId: string
}

export async function getBlockChildrenDeep ({ notionToken, pageId, filter, includeId }: GetBlockChildrenDeepOptions): Promise<NotionBlock[]> {
  const blocks = await getBlockChildren({ notionToken, pageId })
  const filteredBlocks = filter ? blocks.filter(filter) : blocks
  const blocksWithChildren = await Promise.all(filteredBlocks.map(async (block) => {
    if (block.has_children && block.type !== 'child_page') {
      block.children = await getBlockChildrenDeep({ notionToken, pageId: block.id, includeId })
    }

    return block
  }))

  return compact(blocksWithChildren)
}

type Content = BookmarkBlockObjectResponse['bookmark']
| BulletedListItemBlockObjectResponse['bulleted_list_item']
| CalloutBlockObjectResponse['callout']
| ChildDatabaseBlockObjectResponse['child_database']
| ChildPageBlockObjectResponse['child_page']
| ColumnBlockObjectResponse['column']
| ColumnListBlockObjectResponse['column_list']
| DividerBlockObjectResponse['divider']
| EmbedBlockObjectResponse['embed']
| EquationBlockObjectResponse['equation']
| FileBlockObjectResponse['file']
| Heading1BlockObjectResponse['heading_1']
| Heading2BlockObjectResponse['heading_2']
| Heading3BlockObjectResponse['heading_3']
| ImageBlockObjectResponse['image']
| LinkPreviewBlockObjectResponse['link_preview']
| LinkToPageBlockObjectResponse['link_to_page']
| NumberedListItemBlockObjectResponse['numbered_list_item']
| ParagraphBlockObjectResponse['paragraph']
| PdfBlockObjectResponse['pdf']
| QuoteBlockObjectResponse['quote']
| SyncedBlockBlockObjectResponse['synced_block']
| TableBlockObjectResponse['table']
| TableOfContentsBlockObjectResponse['table_of_contents']
| TableRowBlockObjectResponse['table_row']
| TemplateBlockObjectResponse['template']
| ToDoBlockObjectResponse['to_do']
| ToggleBlockObjectResponse['toggle']
| UnsupportedBlockObjectResponse['unsupported']
| VideoBlockObjectResponse['video']

export interface OwnBlock {
  children?: OwnBlock[] | NotionBlock[]
  type: BlockObjectResponse['type']
  content: Content
}

export interface NotionBlock extends OwnBlock {
  children?: NotionBlock[]
  has_children: boolean
  id: string
}

export type Block = OwnBlock | NotionBlock

function getRichText (block: OwnBlock): RichTextItemResponse[] | undefined {
  return 'rich_text' in block.content ? block.content.rich_text : undefined
}

interface MakeAppendRequestOptions {
  blocks: OutgoingBlock[]
  notionToken: string
  pageId: string
}

export function makeAppendRequest ({ notionToken, pageId, blocks }: MakeAppendRequestOptions) {
  return makeRequest<ListBlockChildrenResponse>({
    notionToken,
    method: 'patch',
    path: `blocks/${pageId}/children`,
    body: {
      children: blocks,
    },
  })
  // don't understand why this fails converage
  /* c8 ignore next */
}

interface AppendBlockChildrenOptions {
  blocks: OwnBlock[]
  notionToken: string
  pageId: string
}

export async function appendBlockChildrenDeep ({ notionToken, pageId, blocks }: AppendBlockChildrenOptions) {
  function moveChildren (blocks: NotionBlock[] | OwnBlock[]) {
    return blocks.map((block) => {
      const convertedBlock = convertBlockToOutgoingBlock(block)

      if (!block.children) return convertedBlock

      // @ts-ignore
      convertedBlock[block.type].children = moveChildren(block.children)

      return convertedBlock
    })
  }

  return makeAppendRequest({ notionToken, pageId, blocks: moveChildren(blocks) })
}

export async function appendBlockChildren ({ notionToken, pageId, blocks }: AppendBlockChildrenOptions) {
  let toAppend: OutgoingBlock[] = []

  for (const block of blocks) {
    const convertedBlock = convertBlockToOutgoingBlock(block)

    if (block.children) {
      const children = block.children

      toAppend.push(convertedBlock)

      const { results } = await makeAppendRequest({ notionToken, pageId, blocks: toAppend })
      toAppend = []

      await appendBlockChildren({
        notionToken,
        pageId: results[results.length - 1].id,
        blocks: children,
      })
    } else {
      toAppend.push(convertedBlock)
    }
  }

  if (toAppend.length) {
    return makeAppendRequest({
      notionToken,
      pageId,
      blocks: toAppend,
    })
  }
}

interface UpdateBlockOptions {
  notionToken: string
  blockId: string
  block: Block
}

export function updateBlock ({ notionToken, blockId, block }: UpdateBlockOptions) {
  const convertedBlock = convertBlockToOutgoingBlock(block)

  return makeRequest<void>({
    notionToken,
    method: 'patch',
    path: `blocks/${blockId}`,
    body: convertedBlock,
  })
}

interface UpdatePageOptions {
  notionToken: string
  pageId: string
  properties: {
    [key: string]: {
      date: {
        start: string
      }
    }
  }
}

export function updatePage ({ notionToken, pageId, properties }: UpdatePageOptions) {
  return makeRequest<void>({
    notionToken,
    body: { properties },
    method: 'patch',
    path: `pages/${pageId}`,
  })
}

interface QueryDatabasesOptions {
  notionToken: string
  databaseId: string
}

export function queryDatabases ({ notionToken, databaseId }: QueryDatabasesOptions) {
  return makeRequest<ListDatabasesResponse>({
    notionToken,
    method: 'post',
    path: `databases/${databaseId}/query`,
  })
}

export function convertNotionBlockToOwnBlock (block: BlockObjectResponse): NotionBlock {
  const type = block.type

  return {
    // @ts-ignore
    content: clone(block[type] as Content),
    has_children: block.has_children,
    id: block.id,
    type,
  }
  // don't understand why this fails converage
  /* c8 ignore next */
}

type OutgoingBlock = UpdateBlockBodyParameters & {
  object: string
  type: BlockObjectResponse['type']
}

export function convertBlockToOutgoingBlock (block: Block): OutgoingBlock {
  return {
    object: 'block',
    type: block.type,
    [block.type]: block.content,
  }
  // don't understand why this fails converage
  /* c8 ignore next */
}

interface MakeBlockOptions {
  annotations?: Partial<RichTextItemResponse['annotations']>
  children?: OwnBlock[] | NotionBlock[]
  text: string
  type: BlockObjectResponse['type']
}

export function makeBlock ({ text, type = 'paragraph', annotations, children }: MakeBlockOptions) {
  return {
    type,
    children,
    content: {
      rich_text: [
        {
          type: 'text',
          text: {
            content: text,
          },
          annotations,
          plain_text: text,
        },
      ],
    },
  } as OwnBlock
}

export function richTextToPlainText (richText: RichTextItemResponse[]) {
  return richText
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}

export function getBlockPlainText (block: Block) {
  const richText = getRichText(block)

  if (!richText) return

  return richTextToPlainText(richText)
}

interface RichTextItem {
  text: {
    content: string
    link?: { url: string } | null
  }
  type?: 'text'
  annotations?: {
    bold?: boolean
    italic?: boolean
    strikethrough?: boolean
    underline?: boolean
    code?: boolean
    color?: ApiColor
  }
}

export function makeTextPart (content: string, color?: ApiColor) {
  const textPart = { text: { content } } as RichTextItem

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

export function getMonthNameFromIndex (monthIndex: number, short = false) {
  return dayjs().month(monthIndex).format(short ? 'MMM' : 'MMMM')
}

export function getMonths ({ short }: { short?: boolean } = {}) {
  return Array.from(new Array(12)).map((_, i) => {
    return getMonthNameFromIndex(i, short)
  })
}

/* c8 ignore start */
const getDisplayPrefix = (block: NotionBlock) => {
  switch (block.type) {
    case 'paragraph': return ''
    case 'heading_1': return '# '
    case 'heading_2': return '## '
    case 'heading_3': return '### '
    case 'to_do': return `[${'checked' in block.content && block.content.checked ? '✓' : ' '}] `
    case 'bulleted_list_item': return '• '
    case 'numbered_list_item': return '1. '
    case 'toggle': return '> '
    case 'code': return '<> '
    case 'quote': return '| '
    case 'callout': return '" '
    default: return ''
  }
}

function getDisplayText (block: NotionBlock) {
  if (block.type === 'child_page') {
    return `[${'title' in block.content ? block.content.title : '<untitled>'}]`
  }

  if (block.type === 'divider') {
    return '-----'
  }

  if (getRichText(block)) {
    return `${getDisplayPrefix(block)}${getBlockPlainText(block)}`
  }

  return `(${block.type})`
}

// displays block for logging/debugging purposes
export function displayBlocks (blocks: NotionBlock[]) {
  blocks.forEach((block) => {
    debug(`[${block.id}] ${getDisplayText(block)}${block.has_children ? ' (parent)' : ''}`)
  })
}
/* c8 ignore stop */
