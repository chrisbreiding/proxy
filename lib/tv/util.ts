import type { Method } from 'axios'
import { debug, debugVerbose } from '../util/debug'
import { getEnv } from '../util/env'
import { request } from '../util/network'

const apikey = getEnv('THETVDB_API_KEY')
const pin = getEnv('THETVDB_PIN')
export const baseUrl = 'https://api4.thetvdb.com'

interface LoginResult {
  status: string
  data: {
    token: string
  }
}

async function authenticate () {
  debugVerbose('authenticate')

  try {
    const { data, status } = await request({
      method: 'post',
      url: `${baseUrl}/v4/login`,
      body: { apikey, pin },
    }) as LoginResult

    if (!data?.token) {
      throw new Error(`Could not authenticate with TheTVDB, status: ${status}`)
    }

    return data.token as string
  } catch (error: any) {
    debug('Authentication failed:', error?.stack || error)

    throw error
  }
}

interface MakeRequestOptions {
  method?: Method
  params?: { [key: string]: string }
  path: string
  token?: string
}

export async function makeRequest (options: MakeRequestOptions) {
  const { method, params, path } = options
  const token = options.token || await authenticate()
  const url = `${baseUrl}/v4/${path}`

  debugVerbose('request: %o', { method, url, params, token })

  // TODO: handle error
  const result = await request({
    headers: {
      'Authorization': `Bearer ${token}`,
    },
    method,
    params,
    url,
  })

  result.token = token

  return result
}
