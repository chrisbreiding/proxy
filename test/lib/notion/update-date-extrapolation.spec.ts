import { describe, expect, it } from 'vitest'

import { updateDateExtrapolation } from '../../../lib/notion/update-date-extrapolation'
import { nockGetBlockChildren, notionFixtureContents, snapshotUpdateBlocks } from './util'

describe('lib/notion/update-date-extrapolation', () => {
  it('updates extrapolated dates based on historical and recent dates', async () => {
    nockGetBlockChildren('date-extrapolation-id', { fixture: 'date-extrapolation/categories' })
    nockGetBlockChildren('historical-id', { fixture: 'date-extrapolation/historical-dates' })
    nockGetBlockChildren('recent-id', { fixture: 'date-extrapolation/recent-dates' })
    nockGetBlockChildren('extrapolated-id', { fixture: 'date-extrapolation/extrapolated-dates' })

    const snapshot = snapshotUpdateBlocks([
      'extrapolated-1',
      'extrapolated-2',
      'extrapolated-3',
    ])

    await updateDateExtrapolation({
      notionToken: 'notion-token',
      dateExtrapolationId: 'date-extrapolation-id',
    })

    await snapshot
  })

  it('throws error if no historical toggle', async () => {
    const categories = notionFixtureContents('date-extrapolation/categories')
    categories.results = categories.results.slice(2)

    nockGetBlockChildren('date-extrapolation-id', { reply: categories })
    nockGetBlockChildren('historical-id', { fixture: 'date-extrapolation/historical-dates' })
    nockGetBlockChildren('recent-id', { fixture: 'date-extrapolation/recent-dates' })
    nockGetBlockChildren('extrapolated-id', { fixture: 'date-extrapolation/extrapolated-dates' })

    try {
      await updateDateExtrapolation({
        notionToken: 'notion-token',
        dateExtrapolationId: 'date-extrapolation-id',
      })

      throw new Error('Should have errored')
    } catch (error: any) {
      expect(error.message).toMatchInlineSnapshot(`
        "The following validation error(s) was/were found:
        - Page must have a \\"Historical\\" toggle block with dates"
      `)
    }
  })

  it('throws error if no recent toggle', async () => {
    const categories = notionFixtureContents('date-extrapolation/categories')
    categories.results = [
      ...categories.results.slice(0, 2),
      ...categories.results.slice(3),
    ]

    nockGetBlockChildren('date-extrapolation-id', { reply: categories })
    nockGetBlockChildren('historical-id', { fixture: 'date-extrapolation/historical-dates' })
    nockGetBlockChildren('recent-id', { fixture: 'date-extrapolation/recent-dates' })
    nockGetBlockChildren('extrapolated-id', { fixture: 'date-extrapolation/extrapolated-dates' })

    try {
      await updateDateExtrapolation({
        notionToken: 'notion-token',
        dateExtrapolationId: 'date-extrapolation-id',
      })

      throw new Error('Should have errored')
    } catch (error: any) {
      expect(error.message).toMatchInlineSnapshot(`
        "The following validation error(s) was/were found:
        - Page must have a \\"Recent\\" toggle block with dates"
      `)
    }
  })

  it('throws error if no extrapolated toggle', async () => {
    const categories = notionFixtureContents('date-extrapolation/categories')
    categories.results = [
      ...categories.results.slice(0, 3),
      ...categories.results.slice(4),
    ]

    nockGetBlockChildren('date-extrapolation-id', { reply: categories })
    nockGetBlockChildren('historical-id', { fixture: 'date-extrapolation/historical-dates' })
    nockGetBlockChildren('recent-id', { fixture: 'date-extrapolation/recent-dates' })
    nockGetBlockChildren('extrapolated-id', { fixture: 'date-extrapolation/extrapolated-dates' })

    try {
      await updateDateExtrapolation({
        notionToken: 'notion-token',
        dateExtrapolationId: 'date-extrapolation-id',
      })

      throw new Error('Should have errored')
    } catch (error: any) {
      expect(error.message).toMatchInlineSnapshot(`
        "The following validation error(s) was/were found:
        - Page must have a \\"Extrapolated\\" toggle block with dates"
      `)
    }
  })
})
