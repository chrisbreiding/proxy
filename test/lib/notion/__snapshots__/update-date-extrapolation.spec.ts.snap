// Vitest Snapshot v1

exports[`lib/notion/update-date-extrapolation > updates extrapolated dates based on historical and recent dates 1`] = `
"- 2022-10-16
- 2022-11-18
- 2022-12-21
"
`;
